Software/real_time/ROS_RoboBuggy/src/robobuggy/src/gps/Transistor_GPS_Broadcaster.cpp

```diff
@@ -41,87 +38,118 @@ int Transistor_GPS_Broadcaster::handle_serial_messages()
-                }
-                else
-                {
-                    gps_serial_buffer = "";
-                }
-
-            }
+                if(!parse_gps_string(gps_serial_buffer)) return -1;
         }
     }
 
     return 0;
 }
 
+bool Transistor_GPS_Broadcaster::parse_gps_string(std::string gps_serial_buffer) {
+    std::istringstream ss(gps_serial_buffer);
+    std::string tokens[10];
+
+    for (int i = 0; i < 10; i++)
+    {
+        std::getline(ss, tokens[i], ',');
+    }
+
+    if (!gpgga_valid(tokens))
+    {
+        ROS_ERROR_STREAM("read a GPGGA when it wasn't!");
+        return false;
+    }
+    else if (!gps_locked(tokens))
+    {
+        ROS_ERROR_STREAM("Not locked yet!");
+    }
+    else
+    {
+        double latitude_deg = 41.414938;//convert_to_latitude(tokens[2]);
+        double longitude_deg = 81.86139667;//convert_to_longitude(tokens[4]);
+
+        //if(tokens[5] == "W") longitude_deg *= -1;
+
+        geodesy::UTMPoint utm_point = gps_get_location(latitude_deg, longitude_deg);
+
+        gps_pub.publish(gps_gen_message(latitude_deg, longitude_deg, utm_point));
+    }
+
+    size_t total_bytes_read = tokens->length();
+    if (gps_serial_buffer.length() > total_bytes_read)
+    {
+        gps_serial_buffer = gps_serial_buffer.substr(total_bytes_read);
+    }
+    else
+    {
+        gps_serial_buffer = "";
+    }
+
+    return true;
+}
+
+bool Transistor_GPS_Broadcaster::gpgga_valid(std::string tokens[]) {
+    return tokens[0] == "$GPGGA";
+}
+
+bool Transistor_GPS_Broadcaster::gps_locked(std::string tokens[]) {
+    return atoi(tokens[6].c_str()) != 0;
+}
+
 double Transistor_GPS_Broadcaster::convert_to_latitude(std::string str)
 {
+    if (str.length() == 0)
+    {
+        ROS_ERROR_STREAM("Tried to convert empty latitude string");
+        return 0;
+    }
+
     double degrees = atof(str.substr(0, 2).c_str());
     double minutes = atof(str.substr(2).c_str());
 
-    return degrees + minutes /  60.0;
+    return degrees + (minutes /  60.0);
 }
 
 double Transistor_GPS_Broadcaster::convert_to_longitude(std::string str)
 {
+    if (str.length() == 0)
+    {
+        ROS_ERROR_STREAM("Tried to convert empty longitude string");
+        return 0;
+    }
     double degrees = atof(str.substr(0, 3).c_str());
     double minutes = atof(str.substr(3).c_str());
 
     return degrees + minutes / 60.0;
 }
+
+geodesy::UTMPoint Transistor_GPS_Broadcaster::gps_get_location(double latitude_deg, double longitude_deg) {
+    geographic_msgs::GeoPoint gps_point;
+    gps_point.latitude = latitude_deg;
+    gps_point.longitude = longitude_deg;
+
+    geodesy::UTMPoint utm_point(gps_point);
+
+    return utm_point;
+}
+
+robobuggy::GPS Transistor_GPS_Broadcaster::gps_gen_message(double latitude_deg, double longitude_deg, geodesy::UTMPoint utm_point) {
+    robobuggy::GPS gps_message;
+    gps_message.Lat_deg = static_cast<float>(latitude_deg);
```

Comment:
is the `static_cast` necessary? I was under the impression they were just assignable